File: packages/contracts/v1/enums.go
// Package v1 contains versioned contract types for the trading platform.
package v1

// OrderSide represents the side of an order (buy or sell).
type OrderSide string

const (
	OrderSideBuy  OrderSide = "BUY"
	OrderSideSell OrderSide = "SELL"
)

// OrderMode represents the execution mode of an order.
type OrderMode string

const (
	OrderModeMarket  OrderMode = "MARKET"  // Execute immediately at current price
	OrderModePending OrderMode = "PENDING" // Execute when condition is met
)

// OrderType represents the type of an order.
type OrderType string

const (
	OrderTypeMarket    OrderType = "MARKET"     // Immediate execution at current price
	OrderTypeBuyLimit  OrderType = "BUY_LIMIT"  // Pending: trigger when ask <= limit_price
	OrderTypeSellLimit OrderType = "SELL_LIMIT" // Pending: trigger when bid >= limit_price
	OrderTypeBuyStop   OrderType = "BUY_STOP"   // Pending: trigger when ask >= stop_price
	OrderTypeSellStop  OrderType = "SELL_STOP"  // Pending: trigger when bid <= stop_price

	// Legacy types (deprecated, kept for backward compatibility)
	OrderTypeLimit OrderType = "LIMIT"
	OrderTypeStop  OrderType = "STOP"
)

// IsPending returns true if the order type is a pending order.
func (ot OrderType) IsPending() bool {
	switch ot {
	case OrderTypeBuyLimit, OrderTypeSellLimit, OrderTypeBuyStop, OrderTypeSellStop,
		OrderTypeLimit, OrderTypeStop:
		return true
	default:
		return false
	}
}

// GetMode returns the order mode for this order type.
func (ot OrderType) GetMode() OrderMode {
	if ot.IsPending() {
		return OrderModePending
	}
	return OrderModeMarket
}

// OrderStatus represents the acknowledgment status of an order.
type OrderStatus string

const (
	OrderStatusAccepted OrderStatus = "ACCEPTED"
	OrderStatusRejected OrderStatus = "REJECTED"
)

// ContestPhase represents the current phase of a contest for Kafka events.
// This is a simplified view of the contest lifecycle for event consumers.
type ContestPhase string

const (
	ContestPhaseUpcoming  ContestPhase = "UPCOMING"  // Draft, Scheduled, RegistrationOpen, RegistrationClosed
	ContestPhaseLive      ContestPhase = "LIVE"      // Running
	ContestPhaseFrozen    ContestPhase = "FROZEN"    // Paused, Settling
	ContestPhaseEnded     ContestPhase = "ENDED"     // Completed, Cancelled
	ContestPhaseCancelled ContestPhase = "CANCELLED" // Cancelled (with refunds)
)

// ContestStatus represents the detailed lifecycle status of a contest.
// This matches the database enum and provides finer-grained state tracking.
type ContestStatus string

const (
	// Draft is the initial state when a contest is created but not published.
	ContestStatusDraft ContestStatus = "draft"

	// Scheduled is when the contest is published and accepting registrations.
	ContestStatusScheduled ContestStatus = "scheduled"

	// RegistrationOpen is when registration is actively open.
	ContestStatusRegistrationOpen ContestStatus = "registration_open"

	// RegistrationClosed is when registration deadline passed or contest is full.
	ContestStatusRegistrationClosed ContestStatus = "registration_closed"

	// Running is the active trading period.
	ContestStatusRunning ContestStatus = "running"

	// Paused is when the contest is temporarily frozen (admin action).
	ContestStatusPaused ContestStatus = "paused"

	// Settling is when trading ended and results are being calculated.
	ContestStatusSettling ContestStatus = "settling"

	// Completed is when settlement is done and prizes distributed.
	ContestStatusCompleted ContestStatus = "completed"

	// Cancelled is when the contest was cancelled (refunds issued).
	ContestStatusCancelled ContestStatus = "cancelled"
)

// IsValid returns true if the contest status is a valid value.
func (cs ContestStatus) IsValid() bool {
	switch cs {
	case ContestStatusDraft, ContestStatusScheduled, ContestStatusRegistrationOpen,
		ContestStatusRegistrationClosed, ContestStatusRunning, ContestStatusPaused,
		ContestStatusSettling, ContestStatusCompleted, ContestStatusCancelled:
		return true
	default:
		return false
	}
}

// ToPhase converts a ContestStatus to its corresponding ContestPhase.
func (cs ContestStatus) ToPhase() ContestPhase {
	switch cs {
	case ContestStatusDraft, ContestStatusScheduled, ContestStatusRegistrationOpen, ContestStatusRegistrationClosed:
		return ContestPhaseUpcoming
	case ContestStatusRunning:
		return ContestPhaseLive
	case ContestStatusPaused, ContestStatusSettling:
		return ContestPhaseFrozen
	case ContestStatusCompleted:
		return ContestPhaseEnded
	case ContestStatusCancelled:
		return ContestPhaseCancelled
	default:
		return ContestPhaseUpcoming
	}
}

// IsFinal returns true if the contest status is a final state.
func (cs ContestStatus) IsFinal() bool {
	return cs == ContestStatusCompleted || cs == ContestStatusCancelled
}

// AllowsTrading returns true if the contest status allows trading.
func (cs ContestStatus) AllowsTrading() bool {
	return cs == ContestStatusRunning
}

// AllowsRegistration returns true if the contest status allows registration.
func (cs ContestStatus) AllowsRegistration() bool {
	return cs == ContestStatusScheduled || cs == ContestStatusRegistrationOpen
}

// CancelReason represents the reason for order cancellation.
type CancelReason string

const (
	CancelReasonUserRequested     CancelReason = "user_requested"     // User manually cancelled the order
	CancelReasonContestEnded      CancelReason = "contest_ended"      // Auto-cancelled when contest ends
	CancelReasonInsufficientFunds CancelReason = "insufficient_funds" // Balance check failed
	CancelReasonExpired           CancelReason = "expired"            // Order reached expiration time
)

// AssetClass represents the category of tradable assets.
type AssetClass string

const (
	AssetClassForex  AssetClass = "forex"
	AssetClassCrypto AssetClass = "crypto"
	AssetClassStocks AssetClass = "stocks"
	AssetClassMixed  AssetClass = "mixed"
)

// IsValid returns true if the asset class is a valid value.
func (ac AssetClass) IsValid() bool {
	switch ac {
	case AssetClassForex, AssetClassCrypto, AssetClassStocks, AssetClassMixed:
		return true
	default:
		return false
	}
}

// ContestDurationType represents the duration category of a contest.
type ContestDurationType string

const (
	ContestDurationRush30Min ContestDurationType = "rush_30min"
	ContestDurationHourly    ContestDurationType = "hourly"
	ContestDurationFourHour  ContestDurationType = "four_hour"
	ContestDurationDaily     ContestDurationType = "daily"
	ContestDurationWeekly    ContestDurationType = "weekly"
)

// IsValid returns true if the duration type is a valid value.
func (dt ContestDurationType) IsValid() bool {
	switch dt {
	case ContestDurationRush30Min, ContestDurationHourly, ContestDurationFourHour, ContestDurationDaily, ContestDurationWeekly:
		return true
	default:
		return false
	}
}

// DurationMinutes returns the standard duration in minutes for this type.
func (dt ContestDurationType) DurationMinutes() int {
	switch dt {
	case ContestDurationRush30Min:
		return 30
	case ContestDurationHourly:
		return 60
	case ContestDurationFourHour:
		return 240
	case ContestDurationDaily:
		return 1440
	case ContestDurationWeekly:
		return 10080
	default:
		return 60 // default to hourly
	}
}

// DefaultQtyAllocation returns the maximum trading QTY for this duration type.
// Values match FIXED_PRODUCT_AND_TECHNICAL_POLICIES §5.5 (integer QTY units):
// 30m=5, 1h=10, 4h=10, 1d=20, 1w=20.
func (dt ContestDurationType) DefaultQtyAllocation() int64 {
	switch dt {
	case ContestDurationRush30Min:
		return 5
	case ContestDurationHourly:
		return 10
	case ContestDurationFourHour:
		return 10
	case ContestDurationDaily:
		return 20
	case ContestDurationWeekly:
		return 20
	default:
		return 10
	}
}

// IsAllowedTradingQty reports whether qty is a product-allowed maximum
// trading QTY (custom contests may only use 5, 10, or 20).
func IsAllowedTradingQty(qty int64) bool {
	switch qty {
	case 5, 10, 20:
		return true
	default:
		return false
	}
}

File: packages/contracts/ts/v1/enums.ts
/**
 * Order side (buy or sell).
 */
export type OrderSide = 'BUY' | 'SELL';

/**
 * Order execution mode.
 * - MARKET: Execute immediately at current price
 * - PENDING: Execute when condition is met
 */
export type OrderMode = 'MARKET' | 'PENDING';

/**
 * Order type.
 * - MARKET: Immediate execution at current price
 * - BUY_LIMIT: Pending order, trigger when ask <= limit_price
 * - SELL_LIMIT: Pending order, trigger when bid >= limit_price
 * - BUY_STOP: Pending order, trigger when ask >= stop_price
 * - SELL_STOP: Pending order, trigger when bid <= stop_price
 * - LIMIT/STOP: Legacy types (deprecated)
 */
export type OrderType =
  | 'MARKET'
  | 'BUY_LIMIT'
  | 'SELL_LIMIT'
  | 'BUY_STOP'
  | 'SELL_STOP'
  | 'LIMIT'  // Deprecated
  | 'STOP';  // Deprecated

/**
 * Check if an order type is a pending order.
 */
export function isPendingOrderType(type: OrderType): boolean {
  return type === 'BUY_LIMIT' || type === 'SELL_LIMIT' || type === 'BUY_STOP' || type === 'SELL_STOP';
}

/**
 * Get the order mode for a given order type.
 */
export function getOrderMode(type: OrderType): OrderMode {
  return isPendingOrderType(type) ? 'PENDING' : 'MARKET';
}

/**
 * Order acknowledgment status.
 */
export type OrderStatus = 'ACCEPTED' | 'REJECTED';

/**
 * Contest phase.
 */
export type ContestPhase = 'UPCOMING' | 'LIVE' | 'FROZEN' | 'ENDED';

/**
 * Asset class for contests.
 */
export type AssetClass = 'forex' | 'crypto' | 'stocks' | 'mixed';

/**
 * Check if an asset class is valid.
 */
export function isValidAssetClass(value: string): value is AssetClass {
  return ['forex', 'crypto', 'stocks', 'mixed'].includes(value);
}

/**
 * Contest duration type.
 */
export type ContestDurationType = 'rush_30min' | 'hourly' | 'four_hour' | 'daily' | 'weekly';

/**
 * Check if a duration type is valid.
 */
export function isValidDurationType(value: string): value is ContestDurationType {
  return ['rush_30min', 'hourly', 'four_hour', 'daily', 'weekly'].includes(value);
}

/**
 * Get the duration in minutes for a duration type.
 */
export function getDurationMinutes(type: ContestDurationType): number {
  const durations: Record<ContestDurationType, number> = {
    rush_30min: 30,
    hourly: 60,
    four_hour: 240,
    daily: 1440,
    weekly: 10080,
  };
  return durations[type];
}

/**
 * Get the default QTY allocation for a duration type.
 */
export function getDefaultQtyAllocation(type: ContestDurationType): number {
  // Product policy §5.5 — integer maximum trading QTY by duration.
  const allocations: Record<ContestDurationType, number> = {
    rush_30min: 5,
    hourly: 10,
    four_hour: 10,
    daily: 20,
    weekly: 20,
  };
  return allocations[type];
}

/** Product-allowed maximum trading QTY values for custom contests. */
export function isAllowedTradingQty(qty: number): boolean {
  return qty === 5 || qty === 10 || qty === 20;
}
